File: Library/DSA Reference/Data Structures/3) Sets/sets.cpp
#include <bits/stdc++.h>
using namespace std;

/*
    ============================================================
    SETS CHEAT SHEET — C++20
    ============================================================

    CORE / MUST KNOW
    ----------------
    set<T>:
        unique + sorted
        search/insert/erase: O(log n)

    unordered_set<T>:
        unique + no ordering guarantee
        search/insert/erase: O(1) average, O(n) worst case

    multiset<T>:
        sorted + duplicates allowed
        search/insert/erase: O(log n) for single-element operations;
        operations involving multiple equivalent keys can include
        the number of matching elements.

    EXTRAS / REFERENCE
    ------------------
    bounds, equal_range, set algorithms, custom comparators,
    unordered_set buckets/hash policy, erase_if, merge/extract,
    Counter-like patterns, etc.
*/

template <class T>
void printSet(const T& s) {
    for (const auto& x : s)
        cout << x << ' ';
    cout << '\n';
}

int main() {

    // ============================================================
    // CORE — MUST KNOW
    // ============================================================

    // ------------------------------------------------------------
    // 1. INITIALISATION
    // ------------------------------------------------------------

    set<int> s;
    unordered_set<int> us;
    multiset<int> ms;

    set<int> s1 = {1, 2, 3, 4, 5};
    unordered_set<int> us1 = {1, 2, 3, 4, 5};
    multiset<int> ms1 = {1, 2, 2, 3, 4};

    // From a range:
    vector<int> values = {5, 1, 3, 3, 2};

    set<int> fromRange(values.begin(), values.end());

    // ------------------------------------------------------------
    // 2. INSERT
    // ------------------------------------------------------------

    // set: duplicates ignored
    s.insert(2);
    s.insert(2);
    s.insert(3);
    s.insert(4);

    // The return value is {iterator, inserted}
    auto [itInsert, inserted] = s.insert(5);

    cout << "Inserted 5? " << boolalpha << inserted << '\n';

    // unordered_set
    us.insert(1);
    us.insert(3);
    us.insert(5);
    us.insert(3);                  // duplicate ignored

    // multiset: duplicates allowed
    ms.insert(2);
    ms.insert(2);
    ms.insert(2);
    ms.insert(3);
    ms.insert(4);

    // ------------------------------------------------------------
    // 3. SIZE / EMPTY
    // ------------------------------------------------------------

    cout << "set size: " << s.size() << '\n';       // O(1)
    cout << "us size: " << us.size() << '\n';       // O(1)
    cout << "ms size: " << ms.size() << '\n';       // O(1)

    cout << "set empty: " << s.empty() << '\n';     // O(1)

    // ------------------------------------------------------------
    // 4. MEMBERSHIP
    // ------------------------------------------------------------

    bool inSet = s.count(3) > 0;                  // O(log n)
    bool inUS = us.count(3) > 0;                  // O(1) avg, O(n) worst

    cout << "3 in set: " << inSet << '\n';
    cout << "3 in unordered_set: " << inUS << '\n';

    // C++20 preferred boolean lookup:
    bool contains = s.contains(3);                 // O(log n)
    bool containsUS = us.contains(3);              // O(1) avg, O(n) worst

    // ------------------------------------------------------------
    // 5. FIND
    // ------------------------------------------------------------

    auto it = s.find(3);

    if (it != s.end())
        cout << "Found: " << *it << '\n';

    auto uit = us.find(3);

    if (uit != us.end())
        cout << "Found in unordered_set: " << *uit << '\n';

    // ------------------------------------------------------------
    // 6. ITERATION
    // ------------------------------------------------------------

    cout << "Ordered set: ";
    for (int x : s)
        cout << x << ' ';
    cout << '\n';

    cout << "Unordered set: ";
    for (int x : us)
        cout << x << ' ';
    cout << '\n';

    // std::set traversal is sorted.
    // std::unordered_set traversal has no sorted-order guarantee.

    // ------------------------------------------------------------
    // 7. SMALLEST / LARGEST IN ORDERED SET
    // ------------------------------------------------------------

    if (!s.empty()) {
        int smallest = *s.begin();                  // O(1)
        int largest = *s.rbegin();                 // O(1)

        cout << "Smallest: " << smallest << '\n';
        cout << "Largest: " << largest << '\n';
    }

    // unordered_set has no meaningful smallest/largest by iteration.

    // ------------------------------------------------------------
    // 8. ERASE
    // ------------------------------------------------------------

    // By key:
    size_t erased = s.erase(2);                    // O(log n), 0 or 1

    // By iterator:
    auto eraseIt = s.find(3);
    if (eraseIt != s.end())
        s.erase(eraseIt);                          // amortized O(1)

    // By range [first,last):
    if (!s.empty()) {
        auto first = s.begin();
        auto last = s.end();

        // Example: erase the first element safely
        s.erase(first, next(first));               // amortized O(1) for one node
    }

    // unordered_set:
    us.erase(1);                                   // O(1) avg, O(n) worst

    // multiset:
    ms.erase(2);                                   // Removes ALL 2s

    // To remove only one occurrence:
    auto mit = ms.find(3);
    if (mit != ms.end())
        ms.erase(mit);

    // IMPORTANT:
    // erase(end()) is invalid because end() is not dereferenceable.
    // An empty range erase(end(), end()) is valid, but erasing a
    // single end() iterator is not.

    // ------------------------------------------------------------
    // 9. CLEAR
    // ------------------------------------------------------------

    s.clear();
    us.clear();
    ms.clear();

    // TC: O(n)

    // ------------------------------------------------------------
    // 10. LOWER / UPPER BOUND — ORDERED SETS
    // ------------------------------------------------------------

    set<int> ts = {10, 20, 30, 40, 50};

    auto lb = ts.lower_bound(30);                   // first >= 30
    auto ub = ts.upper_bound(30);                   // first > 30

    if (lb != ts.end())
        cout << "Lower bound: " << *lb << '\n';

    if (ub != ts.end())
        cout << "Upper bound: " << *ub << '\n';

    // Both: O(log n)

    // ------------------------------------------------------------
    // 11. MULTISET
    // ------------------------------------------------------------

    multiset<int> multi = {10, 20, 20, 20, 30, 40};

    cout << "Multiset count(20): "
         << multi.count(20) << '\n';

    auto mlb = multi.lower_bound(20);
    auto mub = multi.upper_bound(20);

    // Range [mlb, mub) contains all 20s.

    if (mlb != multi.end())
        cout << "Multiset lower_bound: " << *mlb << '\n';

    if (mub != multi.end())
        cout << "Multiset upper_bound: " << *mub << '\n';

    // count(x): O(log n + number_of_matches)
    // erase(x): O(log n + number_of_matches)

    // ============================================================
    // EXTRAS — REFERENCE
    // ============================================================

    // ------------------------------------------------------------
    // 12. SET / MULTISET REVERSE TRAVERSAL
    // ------------------------------------------------------------

    set<int> ordered = {1, 2, 3, 4, 5};

    cout << "Reverse ordered set: ";
    for (auto rit = ordered.rbegin();
         rit != ordered.rend();
         ++rit) {
        cout << *rit << ' ';
    }
    cout << '\n';

    // ------------------------------------------------------------
    // 13. EQUAL RANGE
    // ------------------------------------------------------------

    multiset<int> equalSet = {1, 2, 2, 2, 3, 4};

    auto range = equalSet.equal_range(2);

    cout << "All 2s: ";
    for (auto i = range.first; i != range.second; ++i)
        cout << *i << ' ';
    cout << '\n';

    // ------------------------------------------------------------
    // 14. SET ALGORITHMS
    // ------------------------------------------------------------

    set<int> A = {1, 2, 3, 4, 5};
    set<int> B = {3, 4, 5, 6, 7};

    set<int> intersectionSet;
    set<int> unionSet;
    set<int> differenceSet;
    set<int> symmetricDifferenceSet;

    // Intersection A ∩ B
    set_intersection(
        A.begin(), A.end(),
        B.begin(), B.end(),
        inserter(intersectionSet, intersectionSet.begin())
    );

    // Union A ∪ B
    set_union(
        A.begin(), A.end(),
        B.begin(), B.end(),
        inserter(unionSet, unionSet.begin())
    );

    // Difference A - B
    set_difference(
        A.begin(), A.end(),
        B.begin(), B.end(),
        inserter(differenceSet, differenceSet.begin())
    );

    // Symmetric difference (A-B) U (B-A)
    set_symmetric_difference(
        A.begin(), A.end(),
        B.begin(), B.end(),
        inserter(
            symmetricDifferenceSet,
            symmetricDifferenceSet.begin()
        )
    );

    cout << "Intersection: ";
    printSet(intersectionSet);

    cout << "Union: ";
    printSet(unionSet);

    cout << "Difference A-B: ";
    printSet(differenceSet);

    cout << "Symmetric difference: ";
    printSet(symmetricDifferenceSet);

    /*
        These algorithms require sorted/appropriately ordered
        input ranges. std::set naturally supplies sorted ranges.
    */

    // ------------------------------------------------------------
    // 15. SET RELATIONS
    // ------------------------------------------------------------

    // Is A a subset of B?
    bool A_subset_B =
        includes(
            B.begin(), B.end(),
            A.begin(), A.end()
        );

    cout << "A subset of B: "
         << A_subset_B << '\n';

    // TC: O(|A| + |B|) for sorted ranges.

    // ------------------------------------------------------------
    // 16. CUSTOM ORDERING
    // ------------------------------------------------------------

    set<int, greater<int>> descending = {
        1, 2, 3, 4, 5
    };

    cout << "Descending set: ";
    printSet(descending);

    // ------------------------------------------------------------
    // 17. CUSTOM COMPARATOR
    // ------------------------------------------------------------

    struct AbsCompare {
        bool operator()(int a, int b) const {
            if (abs(a) != abs(b))
                return abs(a) < abs(b);

            return a < b;
        }
    };

    set<int, AbsCompare> custom = {
        -4, 2, -2, 1, 4
    };

    cout << "Custom-comparator set: ";
    printSet(custom);

    // IMPORTANT:
    // Set uniqueness is determined by equivalence under Compare,
    // not necessarily operator==.

    // ------------------------------------------------------------
    // 18. COPY / MOVE / SWAP
    // ------------------------------------------------------------

    set<int> original = {1, 2, 3};

    set<int> copied = original;                 // O(n)

    set<int> moved = move(original);             // linear in general
                                                  // source remains valid
                                                  // but unspecified

    set<int> other = {7, 8};
    moved.swap(other);                          // O(1)

    // ------------------------------------------------------------
    // 19. C++20 erase_if
    // ------------------------------------------------------------

    set<int> filterSet = {1, 2, 3, 4, 5, 6};

    erase_if(
        filterSet,
        [](int x) {
            return x % 2 == 0;
        }
    );

    cout << "After erase_if: ";
    printSet(filterSet);

    // ------------------------------------------------------------
    // 20. UNORDERED_SET HASH POLICY
    // ------------------------------------------------------------

    unordered_set<int> hashSet = {
        1, 2, 3, 4, 5
    };

    cout << "Bucket count: "
         << hashSet.bucket_count() << '\n';

    cout << "Load factor: "
         << hashSet.load_factor() << '\n';

    // reserve() requests enough buckets for at least n elements.
    hashSet.reserve(100);

    // rehash() requests at least a given number of buckets.
    hashSet.rehash(50);

    /*
        Core DSA usage:
            unordered_set -> use it for fast average membership.

        You usually do NOT need bucket_count/load_factor/rehash
        in normal DSA problem solving.
    */

    // ------------------------------------------------------------
    // 21. MERGE
    // ------------------------------------------------------------

    set<int> left = {1, 2, 3};
    set<int> right = {3, 4, 5};

    left.merge(right);

    /*
        For std::set:
            elements that do not violate uniqueness are transferred.

        An element already present in left remains in right.
    */

    // ------------------------------------------------------------
    // 22. EXTRACT
    // ------------------------------------------------------------

    set<int> extractSet = {1, 2, 3};

    auto node = extractSet.extract(2);

    if (!node.empty())
        cout << "Extracted: " << node.value() << '\n';

    // node can be modified and inserted into another compatible set.

    // ------------------------------------------------------------
    // 23. PRINT FINAL CONTAINERS
    // ------------------------------------------------------------

    cout << "\nOrdered set: ";
    printSet(ordered);

    cout << "Multiset: ";
    printSet(equalSet);

    cout << "Unordered set: ";
    printSet(hashSet);

    return 0;
}

File: Library/DSA Reference/Data Structures/3) Sets/sets.py
"""
    ============================================================
    SETS CHEAT SHEET — Python 3.x
    ============================================================

    CORE / MUST KNOW
    ----------------
    Main DSA set type: set

    Membership         O(1) average, O(n) worst
    Add                O(1) average, O(n) worst
    Remove             O(1) average, O(n) worst
    Discard            O(1) average, O(n) worst
    Length             O(1)
    Iteration          O(n)

    Python sets:
        - contain unique hashable elements
        - are unordered
        - do not support indexing or slicing

    EXTRAS / REFERENCE
    ------------------
    frozenset, subset/superset/disjoint checks, set comprehensions,
    update variants, Counter as a multiset-like tool, etc.
"""

from collections import Counter


# ================================================================
# CORE — MUST KNOW
# ================================================================

# ------------------------------------------------
# 1. INITIALISATION
# ------------------------------------------------

my_set = {1, 2, 3, 4, 5}

# Constructor:
my_set2 = set([1, 2, 3, 4, 5])

# IMPORTANT:
empty_set = set()

# {} creates an empty DICTIONARY, not an empty set.
empty_dict = {}

fruits = {"apple", "banana", "cherry"}


# ------------------------------------------------
# 2. ADD
# ------------------------------------------------

fruits.add("orange")

# Adding an existing value does nothing:
fruits.add("orange")


# ------------------------------------------------
# 3. MEMBERSHIP
# ------------------------------------------------

if "apple" in fruits:
    print("Apple is in the set.")

present = "banana" in fruits
missing = "mango" not in fruits

# Average TC: O(1)
# Worst TC: O(n)


# ------------------------------------------------
# 4. REMOVE
# ------------------------------------------------

# remove():
# Raises KeyError if the element is absent.

if "banana" in fruits:
    fruits.remove("banana")

# discard():
# Safe if absent; does nothing.

fruits.discard("mango")

# Both: average O(1), worst O(n)


# ------------------------------------------------
# 5. LENGTH / EMPTY
# ------------------------------------------------

n = len(fruits)                # O(1)
is_empty = not fruits          # O(1)

print("Size:", n)
print("Empty:", is_empty)


# ------------------------------------------------
# 6. ITERATION
# ------------------------------------------------

for fruit in fruits:
    print(fruit)

# TC: O(n)

# IMPORTANT:
# No guaranteed sorted order.
# Do not build logic that depends on set iteration order.


# ------------------------------------------------
# 7. CLEAR
# ------------------------------------------------

fruits.clear()

# TC: O(n)


# ------------------------------------------------
# 8. UNION
# ------------------------------------------------

set1 = {1, 2, 3}
set2 = {3, 4, 5}

union_set = set1 | set2

# Method form:
union_set_2 = set1.union(set2)

print("Union:", union_set)

# TC: O(len(set1) + len(set2))


# ------------------------------------------------
# 9. INTERSECTION
# ------------------------------------------------

intersection_set = set1 & set2

# Method:
intersection_set_2 = set1.intersection(set2)

print("Intersection:", intersection_set)

# TC: O(min(len(set1), len(set2)))
# Under normal set/set use.


# ------------------------------------------------
# 10. DIFFERENCE
# ------------------------------------------------

difference_set = set1 - set2

difference_set_2 = set1.difference(set2)

print("Difference:", difference_set)

# Typical TC: O(len(set1))


# ------------------------------------------------
# 11. SYMMETRIC DIFFERENCE
# ------------------------------------------------

sym_diff = set1 ^ set2

sym_diff_2 = set1.symmetric_difference(set2)

print("Symmetric Difference:", sym_diff)

# Typical TC: O(len(set1) + len(set2))


# ================================================================
# EXTRAS — REFERENCE
# ================================================================

# ------------------------------------------------
# 12. SET COMPREHENSION
# ------------------------------------------------

squares = {x * x for x in range(10)}

even_squares = {
    x * x
    for x in range(10)
    if x % 2 == 0
}


# ------------------------------------------------
# 13. UPDATE
# ------------------------------------------------

a = {1, 2, 3}
b = {3, 4, 5}

a.update(b)                  # Mutates a
# a |= b                     # Equivalent operator form

# TC: O(len(b)) average-oriented


# ------------------------------------------------
# 14. INTERSECTION UPDATE
# ------------------------------------------------

a = {1, 2, 3}
b = {2, 3, 4}

a.intersection_update(b)
# a &= b

# Mutates a.


# ------------------------------------------------
# 15. DIFFERENCE UPDATE
# ------------------------------------------------

a = {1, 2, 3}
b = {2, 3, 4}

a.difference_update(b)
# a -= b

# Mutates a.


# ------------------------------------------------
# 16. SYMMETRIC DIFFERENCE UPDATE
# ------------------------------------------------

a = {1, 2, 3}
b = {2, 3, 4}

a.symmetric_difference_update(b)
# a ^= b

# Mutates a.


# ------------------------------------------------
# 17. SUBSET / SUPERSET
# ------------------------------------------------

a = {1, 2}
b = {1, 2, 3, 4}

print(a.issubset(b))
print(a <= b)

print(b.issuperset(a))
print(b >= a)

# Proper subset:
print(a < b)

# Proper superset:
print(b > a)


# ------------------------------------------------
# 18. DISJOINT
# ------------------------------------------------

a = {1, 2}
b = {3, 4}

print(a.isdisjoint(b))
# True

# No common element.


# ------------------------------------------------
# 19. SET COMPARISON
# ------------------------------------------------

a = {1, 2, 3}
b = {3, 2, 1}

print(a == b)                 # True

# Set equality depends on members, not iteration order.

# Set ordering is about subset relationships, not lexicographic
# ordering like strings/lists.


# ------------------------------------------------
# 20. COPY
# ------------------------------------------------

a = {1, 2, 3}

b = a.copy()
c = set(a)

# Shallow copy.
# For immutable/hashable elements, this is normally all you need.

# TC: O(n)


# ------------------------------------------------
# 21. POP
# ------------------------------------------------

a = {1, 2, 3}

x = a.pop()

print("Popped:", x)

# IMPORTANT:
# pop() removes and returns an ARBITRARY element.
# It is NOT "remove the first" or "remove the last".

# Raises KeyError if the set is empty.


# ------------------------------------------------
# 22. HASHABLE ELEMENTS
# ------------------------------------------------

valid = {
    1,
    "hello",
    (1, 2)
}

# Lists are unhashable:
#
# invalid = {[1, 2]}          # TypeError

# Mutable set is also unhashable:
#
# invalid = {{1,2}}            # TypeError

# Use frozenset for a set-like hashable value.


# ------------------------------------------------
# 23. FROZENSET
# ------------------------------------------------

frozen_set = frozenset([1, 2, 3, 4])

# Cannot mutate:
# frozen_set.add(5)            # AttributeError

# Can perform set operations:
other = frozenset([3, 4, 5])

print(frozen_set | other)
print(frozen_set & other)

# frozenset is immutable and hashable.


# ------------------------------------------------
# 24. FROZENSET AS SET ELEMENT
# ------------------------------------------------

nested_sets = {
    frozenset({1, 2}),
    frozenset({3, 4})
}

print(nested_sets)


# ------------------------------------------------
# 25. COUNTER — MULTISET-LIKE DSA TOOL
# ------------------------------------------------

from collections import Counter

# Create Counter
elements = [
    'apple', 'banana', 'apple',
    'orange', 'banana', 'banana'
]

counter = Counter(elements)
print(counter)
# Counter({'banana': 3, 'apple': 2, 'orange': 1})

# Access count
print(counter['apple'])          # 2

# Missing key -> 0
print(counter['grape'])          # 0

# n most common elements
print(counter.most_common(2))
# [('banana', 3), ('apple', 2)]

# Elements repeated according to their counts
print(list(counter.elements()))
# ['apple', 'apple', 'banana', 'banana', 'banana', 'orange']

# Add counts
counter.update(['apple', 'orange', 'orange'])
print(counter)
# Counter({'banana': 3, 'orange': 3, 'apple': 3})

# Subtract counts
counter.subtract(['banana', 'orange'])
print(counter)
# Counter({'banana': 2, 'orange': 2, 'apple': 3})


# ============================================================
# Counter arithmetic
# ============================================================

counter1 = Counter(a=3, b=1)
counter2 = Counter(a=1, b=2)

# Addition → add counts
print(counter1 + counter2)
# Counter({'a': 4, 'b': 3})

# Subtraction → subtract counts, discard zero/negative results
print(counter1 - counter2)
# Counter({'a': 2})

# Intersection → minimum count
print(counter1 & counter2)
# Counter({'a': 1, 'b': 1})

# Union → maximum count
print(counter1 | counter2)
# Counter({'a': 3, 'b': 2})

# Counter is often the Python replacement for:
# C++ multiset / frequency-map style problems,
# depending on what the problem actually needs.


# ------------------------------------------------
# 26. REMOVE DUPLICATES
# ------------------------------------------------

arr = [1, 2, 2, 3, 3, 4]

unique = set(arr)

print("Unique:", unique)

# TC: O(n) average
# SC: O(n)


# Preserve original order while deduplicating:
unique_ordered = list(dict.fromkeys(arr))

print("Unique preserving order:", unique_ordered)

# This is a Python-specific idiom.


# ------------------------------------------------
# 27. SET OPERATIONS WITH ITERABLES
# ------------------------------------------------

a = {1, 2, 3}

# Method forms can accept any iterable:
print(a.union([3, 4, 5]))
print(a.intersection([2, 3, 4]))
print(a.difference([2]))

# Operator forms expect set-like operands:
# a | [3,4]                # TypeError


# ------------------------------------------------
# 28. MANUAL MEMBERSHIP
# ------------------------------------------------

def contains(s, x):
    return x in s

# Average TC: O(1)
# Worst TC: O(n)


# ------------------------------------------------
# 29. SET RELATION QUICK TESTS
# ------------------------------------------------

A = {1, 2, 3}
B = {1, 2, 3, 4, 5}
C = {7, 8}

subset = A <= B
proper_subset = A < B
superset = B >= A
proper_superset = B > A
disjoint = A.isdisjoint(C)

print(subset)
print(proper_subset)
print(superset)
print(proper_superset)
print(disjoint)


# ------------------------------------------------
# 30. CORE COMPLEXITY MEMORY
# ------------------------------------------------

"""
Python set / frozenset:

Membership      O(1) average, O(n) worst
Add             O(1) average, O(n) worst
Remove          O(1) average, O(n) worst
Discard         O(1) average, O(n) worst
Length          O(1)
Iteration       O(n)
Copy            O(n)

Union           O(n+m) average-oriented
Intersection    O(min(n,m)) average-oriented for set/set
Difference      O(n) where n = len(left operand)
Symmetric diff  O(n+m)

No indexing.
No slicing.
No meaningful first/last element.
"""


# ------------------------------------------------
# 31. IMPORTANT PYTHON NOTES
# ------------------------------------------------

"""
1. set stores UNIQUE hashable elements.
2. set is unordered; do not rely on iteration order.
3. {} creates an empty dict, NOT an empty set.
4. Use set() for an empty set.
5. add() inserts one element.
6. remove() raises KeyError if absent.
7. discard() safely removes if present.
8. pop() removes an arbitrary element.
9. set elements must be hashable.
10. frozenset is immutable and hashable.
11. Sets support membership and mathematical set operations.
12. Sets do NOT support indexing or slicing.
"""
